### PR TITLE
[FEMDEM] Improving the coupling and solving errors

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
that's it

### CODE REVIEW

File: applications/FemToDemApplication/custom_elements/generic_small_strain_femdem_element.cpp [hunk 1]
```diff
@@ -457,8 +457,8 @@ void GenericSmallStrainFemDemElement<TDim,TyieldSurf>::IntegrateStressDamageMech
         double damage_parameter; // A parameter
         this->CalculateDamageParameter(rValues, damage_parameter, CharacteristicLength);
         this->CalculateExponentialDamage(rDamage, damage_parameter, uniaxial_stress, initial_threshold);
-        if (this->GetProperties()[FRAGILE])
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/generic_small_strain_femdem_element.cpp]: Remove if unuswd

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+/***********************************************************************************/
+/***********************************************************************************/
+
+void FixFreeVelocityOnNodesProcess::Execute() 
+{
+    const auto& it_node_begin = mrModelPart.NodesBegin();
+    #pragma omp parallel for
+    for (int i = 0; i < static_cast<int>(mrModelPart.Nodes().size()); i++) {
+        auto it_node = it_node_begin + i;
+        if (mFreeOrFix == "Fix") {
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]: Compare strings is expensive, use enums

File: applications/FemToDemApplication/python_scripts/MainFEM_for_coupling.py [hunk 3]
```diff
@@ -31,6 +32,59 @@ def Initialize(self):
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NORMAL)
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.PRESSURE)
 
+        # Adding PFEM Variables TODO put in another place
```
Comment (loumalouomega) [inline applications/FemToDemApplication/python_scripts/MainFEM_for_coupling.py:35]: Pfem uses a lot of historical values, this is super expensive in memory, pfem group should change that

File: applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp [hunk 4]
```diff
@@ -47,589 +47,455 @@ namespace Kratos
...
+            ModelPart::NodesContainerType &rNodes = mrModelPart.Nodes();
+
+            int el = 0;
+            int number = 0;
+
+            //#pragma omp parallel for reduction(+:number) private(el)
+            for (el = 0; el < OutNumberOfElements; el++) {
+                Geometry<Node<3>> vertices;
+
+                unsigned int numfreesurf = 0;
```
Comment (loumalouomega) [inline applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp:150]: This is not efficient, it is not openmp and you are allocating inside the loop

File: applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp [hunk 5]
```diff
@@ -47,589 +47,455 @@ namespace Kratos
...
+
+                if (accepted) {
+                    number += 1;
+                    mrRemesh.PreservedElements[el] = number;
+                }
+            }
+            mrRemesh.Info->NumberOfElements = number;
+        }
+
+        if (mEchoLevel > 1) {
```
Comment (loumalouomega) [inline applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp:361]: Logger

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Some comments

File: applications/FemToDemApplication/custom_elements/generic_small_strain_femdem_element.cpp (hunk 1 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/generic_small_strain_femdem_element.cpp]: done

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]: done

File: applications/FemToDemApplication/python_scripts/MainFEM_for_coupling.py (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/python_scripts/MainFEM_for_coupling.py:35]: I know... I ping @AFranci just to keep him informed

File: applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp:150]: @AFranci maybe we should check this... This process is called at each step so maybe we can reduce a lot of time

**Review by loumalouomega — APPROVED** [review]
OK

File: applications/FemToDemApplication/python_scripts/MainFEM_for_coupling.py (hunk 3 above)
Comment (AFranci) [reply applications/FemToDemApplication/python_scripts/MainFEM_for_coupling.py:35]: Hi @loumalouomega, do you some suggestions to make it better? thanks...

**Pull request merged**